Mathlib/CategoryTheory/Retract.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+
+/-!
+# Retracts
+
+Defines retracts of objects and morphisms.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+/-- An object `X` is a retract of `Y` if there are morphisms `i : X ⟶ Y` and `r : Y ⟶ X` such
+that `i ≫ r = 𝟙 X`. -/
+class Retract (X Y : C) where
```

Comment:
At some point, we may want to add a class `IsRetract` (I am not so sure), but the version with data should be a structure only:
```suggestion
structure Retract (X Y : C) where
```